File: 2023/02_CubeConundrum.py
import sys
import re
import collections

inp = []
with open(sys.argv[1]) as f:
    inp = [l.strip() for l in f.readlines()]

p1 = 0
p2 = 0

for l in inp:
    game, sets = l.split(":")
    _, i = game.split(" ")
    sets = [f.strip().split(" ") for f in re.split(",|;", sets)]

    colors = collections.defaultdict(lambda: [0])

    for a, color in sets:
        colors[color].append(int(a))

    p1 += int(i) if max(colors["red"]) <= 12 and max(colors["green"]) <= 13 and max(colors["blue"]) <= 14 else 0
    p2 += max(colors["red"]) * max(colors["green"])  * max(colors["blue"])

print(p1)
print(p2)

File: 2023/02_CubeConundrum.js
import { readInput, textToArray } from '../util/utils.js';

const list = textToArray(readInput());

const part1 = (list) => {
  const res = list.reduce((acc, [id, colors]) => {
    if (colors["red"] <= 12 && colors["green"] <= 13 && colors["blue"] <= 14) {
      return acc + id;
    }
    return acc;
  }, 0);

  return res;
};

const part2 = (list) => {
  const res = list.reduce((acc, [, colors]) => {
    return acc + (colors["red"] * colors["green"] * colors["blue"]);
  }, 0);

  return res;
};

const parsed = list.map(l => {
  const [game, sets] = l.split(":");
  const [, i] = game.split(" ");

  const colors = sets
    .split(/,|;/)
    .map(f => f.trim().split(" "))
    .reduce((colors, [a, color]) => {
      if (!(color in colors)) colors[color] = [];
      colors[color].push(parseInt(a));
      return colors;
    }, {});

  const maxColors = Object.fromEntries(Object.entries(colors).map(([c, v]) => [c, Math.max(...v)]));

  return [parseInt(i), maxColors];
});

console.log(part1(parsed));
console.log(part2(parsed));
